src/pow.rs

```diff
@@ -0,0 +1,1892 @@
+            }
+        }
+
+        impl Mul<$int> for $ty {
+            type Output = $ty;
+            #[inline]
+            fn mul(self, rhs: $int) -> Self { $ty(self.0.mul(rhs)) }
+        }
+
+        impl Mul<$ty> for $int {
+            type Output = $ty;
+            #[inline]
+            fn mul(self, rhs: $ty) -> $ty { $ty(self.mul(rhs.0)) }
+        }
+    };
+}
+
+// Implement traits and methods shared by `Target` and `Work`.
+macro_rules! do_impl {
+    ($ty:ident) => {
+        impl $ty {
+            /// Creates `Self` from a big-endian array of `u128`s.
+            #[inline]
+            pub fn from_be_array_u128(array: [u128; 2]) -> $ty {
+                $ty(Uint256::from_be_array_u128((array)))
+            }
+
+            /// Creates `Self` from a little-endian array of `u128`s.
+            #[inline]
+            pub fn from_le_array_u128(array: [u128; 2]) -> $ty {
+                $ty(Uint256::from_le_array_u128((array)))
+            }
+
+            /// Creates `Self` from a big-endian array of `u64`s.
+            #[inline]
+            pub fn from_be_array_u64(array: [u64; 4]) -> $ty {
+                $ty(Uint256::from_be_array_u64((array)))
+            }
+
+            /// Creates `Self` from a little-endian array of `u64`s.
+            #[inline]
+            pub fn from_le_array_u64(array: [u64; 4]) -> $ty {
+                $ty(Uint256::from_le_array_u64((array)))
+            }
+
+            /// Creates `Self` from a big-endian array of `u32`s.
+            #[inline]
+            pub fn from_be_array_u32(array: [u32; 8]) -> $ty {
+                $ty(Uint256::from_be_array_u32((array)))
+            }
+
+            /// Creates `Self` from a little-endian array of `u32`s.
+            #[inline]
+            pub fn from_le_array_u32(array: [u32; 8]) -> $ty {
+                $ty(Uint256::from_le_array_u32((array)))
+            }
+
+            /// Creates `Self` from a big-endian array of `u16`s.
+            #[inline]
+            pub fn from_be_array_u16(array: [u16; 16]) -> $ty {
+                $ty(Uint256::from_be_array_u16((array)))
+            }
+
+            /// Creates `Self` from a little-endian array of `u16`s.
+            #[inline]
+            pub fn from_le_array_u16(array: [u16; 16]) -> $ty {
+                $ty(Uint256::from_le_array_u16((array)))
+            }
+
+            /// Creates `Self` from a big-endian array of `u8`s.
+            #[inline]
+            pub fn from_be_array_u8(array: [u8; 32]) -> $ty {
+                $ty(Uint256::from_be_array_u8((array)))
+            }
+
+            /// Creates `Self` from a little-endian array of `u8`s.
+            #[inline]
+            pub fn from_le_array_u8(array: [u8; 32]) -> $ty {
+                $ty(Uint256::from_le_array_u8((array)))
+            }
+
+            /// Creates `Self` from a big-endian array of `u128`s.
+            #[inline]
+            pub fn from_be(array: [u128; 2]) -> $ty { $ty::from_be_array_u128(array) }
```

Comment:
Related to my other comment -- why does the name `from_be` go to the `u128` conversion? Internal storage is an implementation detail, and intuitively I would expect a 256-bit type to be convertible to/from bytes rather than any intermediate-sized integer.